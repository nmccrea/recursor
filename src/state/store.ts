/**
 * Configures and exports the redux store and other top-level Redux resources.
 * @packageDocumentation
 */

import { configureStore, combineReducers } from "@reduxjs/toolkit"
import similarities from "./similarities/reducer"

const rootReducer = combineReducers({ similarities })
const createStore = () => configureStore({ reducer: rootReducer })

export { createStore }
export type RootState = ReturnType<typeof rootReducer>
export type Selector<ReturnType> = (state: RootState) => ReturnType | undefined
