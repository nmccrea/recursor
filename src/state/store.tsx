/**
 * Configures and exports the redux store and other top-level Redux resources.
 * @packageDocumentation
 */

import { combineReducers, configureStore } from "@reduxjs/toolkit"

const rootReducer = combineReducers({})
export type RootState = ReturnType<typeof rootReducer>
export default configureStore({ reducer: rootReducer })
