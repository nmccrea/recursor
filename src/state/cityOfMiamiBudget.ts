/**
 * TODO: type
 * TODO: test
 * TODO: doc
 */

import { RootState } from "./store"
import {
  createAsyncThunk,
  createSlice,
  CaseReducer,
  PayloadAction,
} from "@reduxjs/toolkit"

// State
enum Status {
  Idle = "idle",
  Pending = "pending",
  Fulfilled = "fulfilled",
}

interface CityOfMiamiBudgetState {
  status: Status
  data?: object
}

const INITIAL_STATE: CityOfMiamiBudgetState = { status: Status.Idle }

// Case Reducers
const fetchDataPending: CaseReducer<CityOfMiamiBudgetState, PayloadAction> = (
  previousState
) => ({
  ...previousState,
  status: Status.Pending,
})

const fetchDataFulfilled: CaseReducer<
  CityOfMiamiBudgetState,
  PayloadAction<object>
> = (previousState, action) => ({
  ...previousState,
  status: Status.Fulfilled,
  data: action.payload,
})

// Async Thunks
const fetchData = createAsyncThunk("fetchData", async (url: string) => {
  const response = await fetch(url)
  return await response.json()
})

// Slice
const { reducer } = createSlice({
  name: "cityOfMiamiBudget",
  initialState: INITIAL_STATE,
  reducers: {},
  extraReducers: {
    [fetchData.pending.type]: fetchDataPending,
    [fetchData.fulfilled.type]: fetchDataFulfilled,
  },
})

// Selectors
const getStatus = (state: RootState) =>
  (state.cityOfMiamiBudget && state.cityOfMiamiBudget.status) || undefined

const getData = (state: RootState) =>
  (state.cityOfMiamiBudget && state.cityOfMiamiBudget.data) || undefined

export default reducer
export { fetchData, getStatus, getData, Status }
