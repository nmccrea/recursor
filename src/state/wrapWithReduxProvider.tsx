/**
 * Implements Gatsby's `wrapRootElement` API.
 * @packageDocumentation
 */

import React from "react"
import { Provider as ReduxProvider } from "react-redux"
import store from "./store"

interface Arg {
  element: React.ReactNode
}
export const wrapWithReduxProvider = ({ element }: Arg) => (
  <ReduxProvider store={store}>{element}</ReduxProvider>
)
