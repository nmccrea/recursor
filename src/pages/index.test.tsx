import React from "react"
import { Provider } from "react-redux"
import { render } from "@testing-library/react"
import { createStore, RootState } from "../state/store"
import TreeFractalPage from "./index"

describe("`<TreeFractalPage />`", () => {
  it("renders correctly when there are no similarities", () => {
    const state: RootState = {
      similarities: { ids: [], entities: {} },
    }

    const { container } = render(
      <Provider store={createStore(state)}>
        <TreeFractalPage />
      </Provider>
    )

    expect(container).toMatchSnapshot()
  })

  it("renders correctly when there are existing similarities", () => {
    const state: RootState = {
      similarities: {
        ids: ["test/a", "test/b"],
        entities: {
          "test/a": {
            id: "test/a",
            translation: 0.2298,
            scale: 0.5651,
            angle: 0.1236,
            depth: 2,
            color: "gray",
          },
          "test/b": {
            id: "test/b",
            translation: 0.2134,
            scale: 0.4909,
            angle: 0.0488,
            depth: 2,
            color: "brown",
          },
        },
      },
    }

    const { container } = render(
      <Provider store={createStore(state)}>
        <TreeFractalPage />
      </Provider>
    )

    expect(container).toMatchSnapshot()
  })
})
