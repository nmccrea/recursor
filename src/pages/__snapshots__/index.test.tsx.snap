// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`\`<TreeFractalPage />\` renders correctly when there are existing similarities 1`] = `
<div>
  <div
    style="width: 95vw; height: 70vh; display: flex; justify-content: center; align-items: flex-end;"
  >
    <div
      style="width: 10px; height: 200px; position: relative;"
    >
      <div
        class="tree source depth-1"
        style="position: absolute; inset: 0; background-color: black; transform-origin: center bottom; transform: translateY(-0%) scale(1) rotate(0rad);"
      >
        <div
          class="tree test/a depth-2"
          style="position: absolute; inset: 0; background-color: gray; transform-origin: center bottom; transform: translateY(-22.98%) scale(0.5651) rotate(0.1236rad);"
        />
        <div
          class="tree test/b depth-2"
          style="position: absolute; inset: 0; background-color: brown; transform-origin: center bottom; transform: translateY(-21.34%) scale(0.4909) rotate(0.0488rad);"
        />
      </div>
    </div>
  </div>
  <button>
    Add Similarity
  </button>
  <div
    style="display: flex;"
  >
    <button>
      Remove
    </button>
    <label>
      <p>
        Angle
      </p>
      <input
        max="6.2832"
        min="-6.2832"
        step="0.0001"
        type="range"
        value="0.1236"
      />
      <p>
        Min: 
        -2.00π rad
      </p>
      <p>
        Max: 
        2.00π rad
      </p>
      <p>
        Value: 
        0.04π rad
      </p>
    </label>
    <label>
      <p>
        Scale
      </p>
      <input
        max="2"
        min="0"
        step="0.05"
        type="range"
        value="0.5651"
      />
      <p>
        Min: 
        0.00
      </p>
      <p>
        Max: 
        2.00
      </p>
      <p>
        Value: 
        0.57
      </p>
    </label>
    <label>
      <p>
        Translation
      </p>
      <input
        max="1.5"
        min="0"
        step="0.05"
        type="range"
        value="0.2298"
      />
      <p>
        Min: 
        0.00
      </p>
      <p>
        Max: 
        1.50
      </p>
      <p>
        Value: 
        0.23
      </p>
    </label>
    <label>
      <p>
        Depth
      </p>
      <input
        max="8"
        min="0"
        step="1"
        type="range"
        value="2"
      />
      <p>
        Min: 
        0
      </p>
      <p>
        Max: 
        8
      </p>
      <p>
        Value: 
        2
      </p>
    </label>
  </div>
  <div
    style="display: flex;"
  >
    <button>
      Remove
    </button>
    <label>
      <p>
        Angle
      </p>
      <input
        max="6.2832"
        min="-6.2832"
        step="0.0001"
        type="range"
        value="0.0488"
      />
      <p>
        Min: 
        -2.00π rad
      </p>
      <p>
        Max: 
        2.00π rad
      </p>
      <p>
        Value: 
        0.02π rad
      </p>
    </label>
    <label>
      <p>
        Scale
      </p>
      <input
        max="2"
        min="0"
        step="0.05"
        type="range"
        value="0.4909"
      />
      <p>
        Min: 
        0.00
      </p>
      <p>
        Max: 
        2.00
      </p>
      <p>
        Value: 
        0.49
      </p>
    </label>
    <label>
      <p>
        Translation
      </p>
      <input
        max="1.5"
        min="0"
        step="0.05"
        type="range"
        value="0.2134"
      />
      <p>
        Min: 
        0.00
      </p>
      <p>
        Max: 
        1.50
      </p>
      <p>
        Value: 
        0.21
      </p>
    </label>
    <label>
      <p>
        Depth
      </p>
      <input
        max="8"
        min="0"
        step="1"
        type="range"
        value="2"
      />
      <p>
        Min: 
        0
      </p>
      <p>
        Max: 
        8
      </p>
      <p>
        Value: 
        2
      </p>
    </label>
  </div>
</div>
`;

exports[`\`<TreeFractalPage />\` renders correctly when there are no similarities 1`] = `
<div>
  <div
    style="width: 95vw; height: 70vh; display: flex; justify-content: center; align-items: flex-end;"
  >
    <div
      style="width: 10px; height: 200px; position: relative;"
    >
      <div
        class="tree source depth-1"
        style="position: absolute; inset: 0; background-color: black; transform-origin: center bottom; transform: translateY(-0%) scale(1) rotate(0rad);"
      />
    </div>
  </div>
  <button>
    Add Similarity
  </button>
</div>
`;
